discretize/base/base_mesh.py

```diff
@@ -1880,6 +1880,46 @@ def get_edge_inner_product(
             f"get_edge_inner_product not implemented for {type(self)}"
         )
 
+
+    def get_edge_inner_product_surface(
+        self,
+        model=None,
+        invert_model=False,
+        invert_matrix=False,
+        do_fast=True,
+        **kwargs,
+    ):
+
+        raise NotImplementedError(
+            f"get_edge_inner_product_surface not implemented for {type(self)}"
+        )
+
+    def get_face_inner_product_surface(
+        self,
+        model=None,
+        invert_model=False,
+        invert_matrix=False,
+        do_fast=True,
+        **kwargs,
+    ):
+
+        raise NotImplementedError(
+            f"get_face_inner_product_surface not implemented for {type(self)}"
+        )
+
+    def get_edge_inner_product_line(
+        self,
+        model=None,
+        invert_model=False,
+        invert_matrix=False,
+        do_fast=True,
+        **kwargs,
+    ):
+
```

Comment:
similar to above, you can do this for all mesh types right here:

```
diag(edge_lengths * model)
```

etc.

Reply:
done